[Cai Zhenze] iP

Nova1729 opened the pull request:
# Friday Chatbot 🤖

> *“Your personal assistant for managing tasks effortlessly, certified by Tony Stark, Iron Man himself!”* - Cai Zhenze

## 🚀 Features
- ✅ **Text-based** interface for fast and efficient task management  
- 📅 **Manage deadlines** and events with dates  
- 📖 **Find tasks** quickly with a keyword search  
- 🏆 **Track completed and pending tasks**  
- ⚡ **Fast & easy to use**  

## 📌 How to Use?
1. **Download** the chatbot from [here](https://github.com/Nova1729/ip)  
2. **Run** the application  
3. **Start adding tasks** (todos, deadlines, events)  
4. **Enjoy hassle-free task management!**  

## 📖 Task List Example
Here’s an example of how your task list might look:
1. [T][X] read book
2. [D][X] return book (by: June 6th)
3. [E][ ] project meeting (from: Aug 6th to Aug 7th)
4. [T][ ] join sports club

---

## ✅ Upcoming Features
- [x] **Basic task management**
- [x] **Save tasks to a file**
- [x] **Find tasks by keyword**
- [ ] **Graphical User Interface (GUI)** _(Coming soon!)_
- [ ] **Natural language processing for commands**
- [ ] **Command recommendations**

---

## 🎯 Why Choose Friday?
- ⚡ **Fast**
- 🎯 **Efficient**
- 💡 **Easy to use**
- 💰 **Free**

> **"Stay organized, stay productive!"** 💼✨

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

Commit:
Renamed Duke.java to Friday.java

Commit:
Code now echos commands entered by the user

Commit:
Code can now store whatever text entered by the user and display them back to the user when requested

Commit:
Code now can mark tasks as done

Commit:
Add support for tracking three types of tasks (Todo, Deadline and Event)

Commit:
Automated text UI testing

Commit:
Code now able to handle errors

Commit:
Add support for deleting tasks from the list

Commit:
Use Enumerations

Commit:
Update code to level 7 save

Commit:
Delete Duke.java

Commit:
Merge branch 'branch-Level-7'

Commit:
Teach the chatbot how to understand dates and times.

Commit:
Try stretch goal

Commit:
Update Parser.java, Todo.java, Task.java, Deadline.java, Event.java

Commit:
Handle edge cases when description of a todo is empty, when description of a deadline doesnt have a /by clause, when description of an event doesnt include /from and /to clauses, when mark, unmark and delete dont have task numbers

Commit:
Change file name duke.txt to Friday.txt

Commit:
Put into packages

Commit:
Merge branch 'add-gradle-support'

Commit:
Add Gradle

Commit:
Add JUnit Tests

Commit:
Create JAR file

TanXingZhong commented on src/main/java/chatbot/Parser.java:
```diff
@@ -0,0 +1,81 @@
+        }
+    }
+
+    public static Task parseTask(String line) throws IllegalArgumentException {
```
I like how you broke down the parser into two parts. 👍

TanXingZhong commented on src/main/java/chatbot/commands/AddDeadlineCommand.java:
```diff
@@ -0,0 +1,29 @@
+        if (!input.contains(" /by ")) {
+            throw new DeadlineException("The description of a deadline must include a /by clause.");
+        }
+        String[] details = input.split(" /by ", 2);
```
Perhaps its better to split and use trim() ? 🤔

TanXingZhong commented on src/main/java/chatbot/commands/AddEventCommand.java:
```diff
@@ -0,0 +1,29 @@
+        if (!input.contains(" /from ") || !input.contains(" /to ")) {
+            throw new EventException("The description of an event must include /from and /to clauses.");
+        }
+        String[] details = input.split(" /from | /to ", 3);
```
Perhaps its better to split and use trim() ? 🤔

TanXingZhong commented on src/main/java/chatbot/commands/CommandType.java:
```diff
@@ -0,0 +1,16 @@
+import chatbot.tasks.*;
+
+public enum CommandType {
+    BYE, LIST, MARK, UNMARK, TODO, DEADLINE, EVENT, DELETE, UNKNOWN;
```
I like how you used enums for commands. 👍

TanXingZhong commented on src/main/java/chatbot/tasks/Deadline.java:
```diff
@@ -0,0 +1,37 @@
+
+    public Deadline(String description, String by) {
+        super(description);
+        this.by = LocalDate.parse(by); // Parse string to LocalDate
```
Perhaps its a good idea to check if by is inthe correct format for localdate? 🤔

TanXingZhong reviewed (commented):
LGTM!!! Just some minor nits 😄

Commit:
Add JavaDoc for FileManager

Commit:
Add JavaDoc for Friday

Commit:
Add JavaDoc for Parser

Commit:
Add Java Doc for Storage

Commit:
Add Java doc for Ui

Commit:
Add Java Doc for Command

Commit:
Add Java Doc for AddDeadlineCommand

Commit:
Add Java Doc for AddEventCommand

Commit:
Add TodoCommand

Commit:
Add Java Doc for CommandType

Commit:
Delete unnecessary code in Friday

Commit:
Add Java Doc for DeleteCommand

Commit:
Add JavaDoc for ExitCommand

Commit:
Add Java Doc for ListCommand

Commit:
Add Java Doc for MarkCommand

Commit:
Add java Doc for UnmarkCommand

Commit:
Add JavaDoc for Task

Commit:
Follow the Coding Standard

Commit:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	CS2103T/CS2103T_chatBot_Tasks/Friday.txt
#	src/main/java/chatbot/Parser.java
#	src/main/java/chatbot/commands/CommandType.java

Commit:
Debug Friday.txt

Commit:
Remove unnecessary code to prevent potential conflicts during merging

Commit:
Give users a way to find a task by searching for a keyword in the task description.

Commit:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/main/java/chatbot/FileManager.java
#	src/main/java/chatbot/Parser.java
#	src/main/java/chatbot/commands/CommandType.java
#	src/main/java/chatbot/tasks/Task.java

Commit:
Solve checkstyle

Commit:
Add new dependencies to gradle

Commit:
Add basic GUI features

Commit:
Java Fx tutorial part 3

Commit:
Update ListCommand.java, MarkCommand.java and AddTodoCommand.java such that the output is shown in the GUI

Commit:
Update UnmarkCommand.java to produce output on GUI

Commit:
Update FindCommand.java to produce output in GUI

Commit:
Update ExitCommand to produce output on GUI

Commit:
Update DeleteCommand.java to produce output on GUI

Commit:
Update AddEventCommand.java to show output in GUI

Commit:
Update AddDeadlineCommand.java to show output on GUI

Commit:
Update code to correctly handle bye command

Commit:
Handle edge cases of delete command

TrKNguyen commented on src/main/java/chatbot/commands/AddDeadlineCommand.java:
Create multiple files for some specific task is very good practice, nice!

TrKNguyen commented on src/main/java/chatbot/Friday.java:
Haven't seen the GUI, hope you will add it soon, Instead of returning a String, I think if you want to capture the output, you can do like capture the output of System.out.print into a String, you can use ByteArrayOutputStream and PrintStream to redirect System. out

TrKNguyen commented on src/main/java/chatbot/FileManager.java:
```diff
@@ -0,0 +1,123 @@
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
```
Uses Modern Java NIO, good

TrKNguyen commented on src/main/java/chatbot/FileManager.java:
Maybe you can use Logger Instead of System.out.println() for Errors

TrKNguyen commented on src/main/java/chatbot/Ui.java:
Scanner is never closed, which can cause resource leaks when used in larger applications.

TrKNguyen reviewed (commented):
Good overall, hope you will add GUI soon

Commit:
Use Scene Builder to tweak the GUI

Commit:
Update code to be able to add multiple deadlines to the list all at once. Each deadline must be separated by "; "

Commit:
Update code to be able to add multiple events to the list all at once. Update Parser.java to only check if the descriptions of deadline and event are empty, delegate rest of the checks to AddEventCommand.java and AddDeadlineCommand.java

Commit:
Update DeadlineException messages

Commit:
Delegate complicated deadline and event checks to new classes checkDeadline.java and checkEvent.java

Commit:
Made DeleteCommand.java more modular

Commit:
Make MarkCommand more modular

Commit:
Make UnmarkCommand more modular

Commit:
Delete unnecessary FileManager.java

Commit:
Add assertions for classes in the check package

Commit:
Add assertions for classes in the commands package

Commit:
Refactor command classes for modularity and separation of concerns

The response generation logic in `AddTodoCommand`, `AddEventCommand`,
and `AddDeadlineCommand` was tightly coupled with the execution logic.
This made the command classes less maintainable and harder to test.

Having response generation in command execution violates the
single responsibility principle (SRP) and makes future modifications
more error-prone.

Let's refactor these command classes to delegate response generation
to dedicated response classes:
- Extracted response logic to:
  - `AddTodoResponse.java`
  - `AddEventResponse.java`
  - `AddDeadlineResponse.java`
- Updated the command classes to use the new response classes.
- Improved readability, testability, and maintainability.

Now, commands focus only on executing logic, while responses are
handled separately, leading to cleaner and more modular code.

Commit:
Commit Message:
Remove unnecessary Ui class and refactor usages to support JavaFX GUI
The Ui class was responsible for handling CLI-based interactions, including reading user input, displaying messages, and handling errors. However, since the application now uses JavaFX for the GUI, the Ui class is redundant.

Why this change is necessary?
The UI logic has been fully transitioned to JavaFX.
The Ui class and its methods (showWelcome, showError, showMessage, etc.) are no longer needed.
The chatbot should no longer rely on Ui to handle interactions.
What has been done?
Deleted Ui.java since it is no longer required.
Removed all references to Ui in:
Friday.java
Parser.java
AddDeadlineCommand.java
FridayTest.java
UiTest.java
Updated error handling to use System.out.println() for now, where necessary.
Modified Friday.java to remove any calls to Ui and ensure commands still work with JavaFX.
Why is it done this way?
This keeps the chatbot cleaner and more modular, fully leveraging JavaFX for UI interactions.
Removing Ui improves maintainability by eliminating an unnecessary abstraction.
Additional Notes
Any remaining System.out.println() calls for errors will later be integrated into proper JavaFX error popups.

Commit:
Refactor FindCommand to use Streams API for better readability and efficiency
Refactored the FindCommand class to replace explicit loops and manual string concatenation with Java Streams API.

Current Issues (Before Refactoring)
Used a manual loop to iterate over tasks and filter matching tasks.
Created a StringBuilder and manually concatenated strings for output.
Used indexed loops for formatting instead of a more declarative approach.
Changes Made
Replaced the loop-based filtering with .stream().filter().collect(Collectors.toList()).
Replaced the manual StringBuilder logic with .stream().map().collect(Collectors.joining()) to format output.
Improved readability and reduced boilerplate code.
Why This Change?
Better Readability: The new implementation is more declarative and easier to understand.
Performance Optimization: Streams allow for efficient filtering and transformation of data.
Concise Code: Eliminates unnecessary loops and manual concatenation.

Commit:
Remove unnecessary FileManager class

Why this change?
The FileManager class was redundant and no longer required for the chatbot's functionality.
Its responsibilities have been absorbed by Storage, making it unnecessary to keep a separate class.
Removing this class simplifies the codebase, reducing maintenance overhead.

Changes Made:
Deleted the FileManager class from the project.
Removed all references to FileManager from other parts of the code.
Ensured that Storage handles all file operations previously managed by FileManager.

Benefits of this refactor:
Code Simplification: Fewer classes mean easier maintenance.
Better Responsibility Allocation: Storage is now the single responsible class for file operations.
Reduced Redundancy: No unnecessary duplication of file handling logic.

Commit:
Update JavaDoc

Nova1729 committed:
Merge pull request #1 from Nova1729/branch-A-Assertions

Branch a assertions

Commit:
Merge branch 'master' of https://github.com/Nova1729/ip

* 'master' of https://github.com/Nova1729/ip:
  Add assertions for classes in the commands package
  Add assertions for classes in the check package
  Delete unnecessary FileManager.java

Commit:
Merge branch 'master' into branch-A-CodeQuality

* master:
  Add assertions for classes in the commands package
  Add assertions for classes in the check package
  Delete unnecessary FileManager.java

# Conflicts:
#	path/to/Friday.txt
#	src/main/java/chatbot/Parser.java
#	src/main/java/chatbot/commands/AddDeadlineCommand.java
#	src/main/java/chatbot/commands/AddEventCommand.java
#	src/main/java/chatbot/commands/AddTodoCommand.java
#	src/main/java/chatbot/commands/DeleteCommand.java
#	src/main/java/chatbot/commands/FindCommand.java
#	src/main/java/chatbot/commands/ListCommand.java
#	src/main/java/chatbot/commands/MarkCommand.java
#	src/main/java/chatbot/commands/UnmarkCommand.java

Commit:
Merge branch 'master' into branch-A-Streams

* master:
  Add assertions for classes in the commands package
  Add assertions for classes in the check package
  Delete unnecessary FileManager.java

# Conflicts:
#	src/main/java/chatbot/commands/FindCommand.java

Nova1729 committed:
Merge pull request #2 from Nova1729/branch-A-CodeQuality

Branch a code quality

Commit:
Merge branch 'master' of https://github.com/Nova1729/ip

* 'master' of https://github.com/Nova1729/ip:
  Update JavaDoc
  Remove unnecessary FileManager class
  Commit Message: Remove unnecessary Ui class and refactor usages to support JavaFX GUI The Ui class was responsible for handling CLI-based interactions, including reading user input, displaying messages, and handling errors. However, since the application now uses JavaFX for the GUI, the Ui class is redundant.
  Refactor command classes for modularity and separation of concerns

Nova1729 committed:
Merge branch 'master' into branch-A-Streams

Nova1729 committed:
Merge pull request #3 from Nova1729/branch-A-Streams

branch a Streams

Commit:
Merge branch 'master' of https://github.com/Nova1729/ip

* 'master' of https://github.com/Nova1729/ip:
  Refactor FindCommand to use Streams API for better readability and efficiency Refactored the FindCommand class to replace explicit loops and manual string concatenation with Java Streams API.

Commit:
Improve code quality

Commit:
Merge branch 'branch-A-Streams' of https://github.com/Nova1729/ip into branch-A-Streams

* 'branch-A-Streams' of https://github.com/Nova1729/ip:
  Update JavaDoc
  Remove unnecessary FileManager class
  Commit Message: Remove unnecessary Ui class and refactor usages to support JavaFX GUI The Ui class was responsible for handling CLI-based interactions, including reading user input, displaying messages, and handling errors. However, since the application now uses JavaFX for the GUI, the Ui class is redundant.
  Refactor command classes for modularity and separation of concerns

Commit:
Merge branch 'branch-A-Streams'

* branch-A-Streams:

Commit:
Update code to check deadline duplicates

Commit:
Update code to handle duplicate event commands

Commit:
Update code to handle todo duplicates

Commit:
Used chatGPT to enhance quality of current code

Commit:
Refactor package names to follow Java naming conventions

Changed all package names to lowercase to align with Java's standard
naming convention. This improves consistency, readability, and
maintainability.

Following Java's package naming conventions ensures better
compatibility with industry standards and avoids potential issues
in case-sensitive file systems.

Commit:
Refactor class names to follow coding standards

Updated class names to align with Clean Code principles and Java
naming conventions. Classes now use nouns instead of verbs, making
their purpose clearer and improving maintainability.

Examples of changes made:
- Renamed `CheckDeadline` to `DeadlineValidator`
- Renamed `CheckEvent` to `EventValidator`
- Renamed `CheckMark` to `MarkValidator`
- Renamed `CheckUnmark` to `UnmarkValidator`
- Renamed `CheckDeadlineDuplicates` to `DuplicateDeadlineChecker`
- Renamed `CheckEventDuplicates` to `DuplicateEventChecker`
- Renamed `CheckTodoDuplicates` to `DuplicateTodoChecker`

These changes enhance code readability and ensure consistency
across the project.

Commit:
Enhance error handling for invalid delete, mark, and unmark inputs

Added improved error handling to validate user input for the delete, mark,
and unmark commands. The system now checks if the input is a valid number
and within the acceptable range of task indices.

Changes include:
- Handling non-numeric inputs by throwing appropriate exceptions.
- Checking if the task index is out of range and providing clear error messages.

This improves user experience by giving informative feedback and preventing
invalid operations.

Commit:
Create website for Friday chatbot

Commit:
Update code

Commit:
Update README.md

Commit:
Update code
